### PR TITLE
Remove the cached immutable type from compile-time BTypes and use a registry to maintain the corresponding immutable types

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title. Making this a draft PR until the full build can be tested.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/35949

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -6563,6 +6570,49 @@ boolean isSameSingletonType(BFiniteType type1, BFiniteType type2) {
...
+    }
+
+    public static Optional<BIntersectionType> getImmutableType(SymbolTable symTable, PackageID packageId,
+                                                               SelectivelyImmutableReferenceType type) {
+        Map<String, Map<SelectivelyImmutableReferenceType, BIntersectionType>> immutableTypeMaps =
+                symTable.immutableTypeMaps;
+
+        String packageIdString = getPackageIdString(packageId);
+
+        if (!immutableTypeMaps.containsKey(packageIdString)) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:6599]: can simplify here as well

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -6563,6 +6570,49 @@ boolean isSameSingletonType(BFiniteType type1, BFiniteType type2) {
...
+    public static void addImmutableType(SymbolTable symTable, PackageID packageId,
+                                        SelectivelyImmutableReferenceType type, BIntersectionType immutableType) {
+        Map<SelectivelyImmutableReferenceType, BIntersectionType> moduleImmutableTypeMap;
+
+        Map<String, Map<SelectivelyImmutableReferenceType, BIntersectionType>> immutableTypeMaps =
+                symTable.immutableTypeMaps;
+
+        String packageIdString = getPackageIdString(packageId);
+
+        if (immutableTypeMaps.containsKey(packageIdString)) {
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed in [926b99f](https://github.com/ballerina-platform/ballerina-lang/pull/36379/commits/926b99fdcf6a549e957df366658d80a0be2fd167).

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -6563,6 +6570,50 @@ boolean isSameSingletonType(BFiniteType type1, BFiniteType type2) {
...
+        Map<SelectivelyImmutableReferenceType, BIntersectionType> moduleImmutableTypeMap =
+                immutableTypeMaps.get(packageIdString);
+
+        if (moduleImmutableTypeMap == null) {
+            return Optional.empty();
+        }
+
+        if (moduleImmutableTypeMap.containsKey(type)) {
+            return Optional.of(moduleImmutableTypeMap.get(type));
+        }
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:6608]:
```suggestion
        if (moduleImmutableTypeMap != null && moduleImmutableTypeMap.containsKey(type)) {
            return Optional.of(moduleImmutableTypeMap.get(type));
        }
```

**Review by pcnfernando — APPROVED** [review]
LGTM

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**